File: dotnet/DictaSqlite/DictaSqlite.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Dicta;
using System.IO;
using System.Data.SQLite;

namespace DictaSqlite
{
    public class DictaSqlite
    {
        private static DictaModel model;

        private static Func<object, object> sqliteQuery = ((dynamic id) =>
        {
            SQLiteConnection connection = new SQLiteConnection(@"Data Source=sqlite\dicta.sqlite;");
            connection.Open();
            SQLiteCommand command = connection.CreateCommand();
            command.CommandText = "SELECT name FROM numbers WHERE id=" + id;
            SQLiteDataReader reader = command.ExecuteReader();
            object name = "";
            while (reader.Read())
            {
                name = reader["name"];
                break;
            }
            connection.Close();
            return name;
        });

        public static void WasMain(string[] args)
        {
            string fileName = @"dicta\sqlite.dicta";
            string text = File.ReadAllText(fileName);
            model = new DictaModel();
            model.Parse(text);
            model.AddFunction("sqlite", sqliteQuery);
            while (true)
            {
                int id;
                Int32.TryParse(Console.ReadLine(), out id);
                if (id != null)
                {
                    model.Set("id", id);
                    var name = model.Get("name");
                    Console.WriteLine(name);
                }
            }
        }

        public static void Main(string[] args)
        {
            string fileName = @"dicta\setget.dicta";
            string text = File.ReadAllText(fileName);
            model = new DictaModel();
            model.Parse(text);
            Random random = new Random();
//            model.AddFunction("sqlite", sqliteQuery);
            int queryCount = (int)1e4;
            DateTime start = DateTime.Now;
            for (int index = 0; index < queryCount; index++)
            {
                int idValue = random.Next(10) + 1;
                model.Set("id", idValue);
                model.Get("name");
            }
            TimeSpan elapsed = DateTime.Now - start;
            double rate = queryCount / elapsed.TotalSeconds;
            Console.Out.WriteLine("{0} queries/second", rate);
            Console.ReadKey();
        }
    }
}

File: java/DictaSqlite/src/net/geocentral/dicta/DictaSqlite.java
package net.geocentral.dicta;

import java.io.File;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.Date;
import java.util.Random;
import java.util.Scanner;

public class DictaSqlite {

    public String sqliteQuery(int id) throws Exception {
        Class.forName("org.sqlite.JDBC");
        Connection connection = DriverManager.getConnection("jdbc:sqlite:sqlite/dicta.sqlite");
        Statement statement = connection.createStatement();
        ResultSet rs = statement.executeQuery("SELECT name FROM numbers WHERE id=" + id);
        String name = "";
        while (rs.next()) {
            name = rs.getString("name");
            break;
        }
        connection.close();
        return name;
    }

    public static void main(String[] args) throws Exception {
        String text = new Scanner(new File("dicta/setget.dicta")).useDelimiter("\\A").next();
        DictaModel model = new DictaModel();
        model.parse(text);
        Random random = new Random();
        int queryCount = (int)1e4;
        Date start = new Date();
        for (int index = 0; index < queryCount; index++) {
            int idValue = random.nextInt(10) + 1;
            model.set("id", idValue);
            model.get("name");
        }
        long elapsed = new Date().getTime() - start.getTime();
        double rate = 1e3*queryCount/elapsed;
        System.out.println(String.format("%s queries/second", rate));
    }

//    public static void main(String[] args) throws Exception {
//        DictaSqlite dictaSqlite = new DictaSqlite();
//        String text = new Scanner(new File("dicta/sqlite.dicta")).useDelimiter("\\A").next();
//        DictaModel model = new DictaModel();
//        model.parse(text);
//        model.AddFunction("sqlite", dictaSqlite, "sqliteQuery");
//        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
//        while (true) {
//            String line = reader.readLine();
//            int id = Integer.valueOf(line);
//            model.set("id", id);
//            Object name = model.get("name");
//            System.out.println(name);
//        }
//    }
}
